crates/wdl-engine/src/cache/hash.rs

```diff
@@ -0,0 +1,320 @@
+use crate::Map;
+use crate::Object;
+use crate::OutputValue;
+use crate::Pair;
+use crate::PrimitiveValue;
+use crate::Struct;
+use crate::Value;
+use crate::digest::Digest;
+use crate::path::EvaluationPath;
+
+/// The variant tag for `None` values.
+const NONE_VARIANT_TAG: u8 = 0;
+/// The variant tag for `Boolean` values.
+const BOOLEAN_VARIANT_TAG: u8 = 1;
+/// The variant tag for `Int` values.
+const INTEGER_VARIANT_TAG: u8 = 2;
+/// The variant tag for `Float` values.
+const FLOAT_VARIANT_TAG: u8 = 3;
+/// The variant tag for `String` values.
+const STRING_VARIANT_TAG: u8 = 4;
+/// The variant tag for `File` values.
+const FILE_VARIANT_TAG: u8 = 5;
+/// The variant tag for `Directory` values.
+const DIRECTORY_VARIANT_TAG: u8 = 6;
+/// The variant tag for `Pair` values.
+const PAIR_VARIANT_TAG: u8 = 7;
+/// The variant tag for `Array` values.
+const ARRAY_VARIANT_TAG: u8 = 8;
+/// The variant tag for `Map` values.
+const MAP_VARIANT_TAG: u8 = 9;
+/// The variant tag for `Object` values.
+const OBJECT_VARIANT_TAG: u8 = 10;
+/// The variant tag for `Struct` values.
+const STRUCT_VARIANT_TAG: u8 = 11;
+/// The variant tag for `hints` values.
+const HINTS_VARIANT_TAG: u8 = 12;
+/// The variant tag for `input` values.
+const INPUT_VARIANT_TAG: u8 = 13;
+/// The variant tag for `output` values.
+const OUTPUT_VARIANT_TAG: u8 = 14;
+
+/// The variant tag for local evaluation paths.
+const LOCAL_PATH_VARIANT_TAG: u8 = 0;
+/// The variant tag for remote evaluation paths.
+const REMOTE_PATH_VARIANT_TAG: u8 = 1;
+
+/// The variant tag for hash-based remote content digests.
+const CONTENT_HASH_VARIANT_TAG: u8 = 0;
+/// The variant tag for ETag based remote content digests.
+const CONTENT_ETAG_VARIANT_TAG: u8 = 1;
+
+/// The variant tag for file digests.
+const DIGEST_FILE_VARIANT_TAG: u8 = 0;
+/// The variant tag for directory digests.
+const DIGEST_DIRECTORY_VARIANT_TAG: u8 = 1;
+
+/// Hashes a sequence of hashable items.
+fn hash_sequence<'a, T: Hashable + 'a>(
+    hasher: &mut Hasher,
+    items: impl ExactSizeIterator<Item = T>,
+) {
+    hasher.update(&(items.len() as u32).to_le_bytes());
+    for item in items {
+        item.hash(hasher);
+    }
+}
+
+/// Trait used to implement WDL value hashing for call caching.
+pub trait Hashable {
+    /// Hashes into the given Blake3 hasher.
+    fn hash(&self, hasher: &mut Hasher);
+}
+
+impl Hashable for &str {
+    fn hash(&self, hasher: &mut Hasher) {
+        hasher.update(&(self.len() as u32).to_le_bytes());
+        hasher.update(self.as_bytes());
+    }
+}
+
+impl Hashable for String {
+    fn hash(&self, hasher: &mut Hasher) {
+        self.as_str().hash(hasher);
+    }
+}
+
+impl Hashable for &[u8] {
+    fn hash(&self, hasher: &mut Hasher) {
+        hasher.update(self);
+    }
+}
+
+impl Hashable for Path {
+    fn hash(&self, hasher: &mut Hasher) {
+        hasher.update(&(self.as_os_str().len() as u32).to_le_bytes());
+        hasher.update(self.as_os_str().as_encoded_bytes());
```

Comment:
This does not match the behavior specified in the RFC that `File`/`Directory` values are hashed as UTF-8, as [`as_encoded_bytes()`](https://doc.rust-lang.org/std/ffi/struct.OsStr.html#method.as_encoded_bytes) does not necessarily return UTF-8. The doc even says:
> sending the slice over the network or storing it in a file will likely result in incompatible byte slices

We should probably use [`to_string_lossy()`](https://doc.rust-lang.org/std/ffi/struct.OsStr.html#method.to_string_lossy) or something similar that checks the UTF-8 validity of its result.